### PR TITLE
feat: training and customisable observations

### PR DESCRIPTION
**Pull request by zombie-einstein**
Hi @sash-a , just making this smaller PR for the training network additions, just wanted your input on the structure and some of the proposed changes, see comments below. This probably still needs some tweaking though.

### CODE REVIEW

File: jumanji/wrappers.py [hunk 1]
```diff
@@ -226,19 +226,23 @@ class MultiToSingleWrapper(
...
             env: `Environment` to wrap to a `dm_env.Environment`.
+            action_shaper: a function that reshapes flattened actions where the
+                environment expects actions shaped for multiple agents.
             reward_aggregator: a function to aggregate all agents rewards into a single scalar
                 value, e.g. sum.
             discount_aggregator: a function to aggregate all agents discounts into a single
                 scalar value, e.g. max.
         """
         super().__init__(env)
+        self.action_shaper = action_shaper
```
Comment (zombie-einstein) [inline jumanji/wrappers.py:245]: I added this action shaper to reshape flattened actions back into a shape that has a dimension for multiple agents, in this case from network output `[n-agents * 2,]` to `[n-agents. 2]`.

In principle the normal distribution should just as well work for a multi-dimensional so the output of the network could be the expected shape, but it seemed like other parts of the code expected network outputs to be 1d array (except in some special cases). Let me know if you think this change makes sense

**Comment by sash-a** [comment]
Thanks for this I'll have a look on Monday/Tuesday. 

I didn't realise we didn't have anything setup for continuous distributions, sorry about that! At a glance your implementation looks pretty great though so thanks for all the extra work!

On a side note I'm surprised we don't use distrax or tfp I wonder how grad will work because they have stuff like `inverse` and `log_det_jacobian`. I guess we'll find out when we train it 😂. I'm not opposed to adding a distribution library as a dependency if needed, but you don't need to do unless something isn't working

**Comment by zombie-einstein** [comment]
> On a side note I'm surprised we don't use distrax or tfp I wonder how grad will work because they have stuff like `inverse` and `log_det_jacobian`. I guess we'll find out when we train it 😂. I'm not opposed to adding a distribution library as a dependency if needed, but you don't need to do unless something isn't working

Haha, yeah actually was going to ask if Distrax was an option but didn't want to make too massive a change. For now I could add it as a dependency, then use it inside the normal distribution so it's relatively low impact?

File: jumanji/training/networks/search_and_rescue/actor_critic.py [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+
+def make_actor_network(layers: Sequence[int], n_actions: int) -> FeedForwardNetwork:
+    def network_fn(observation: Observation) -> Union[chex.Array, Tuple[chex.Array, chex.Array]]:
+        views = observation.searcher_views  # (B, M, V)
+        x = views.reshape(views.shape[0], -1)  # (B, N)
+        means = hk.nets.MLP([*layers, n_actions])(x)  # (B, A)
+        log_stds = hk.get_parameter(
+            "log_stds", shape=means.shape[1:], init=hk.initializers.Constant(0.1)
+        )  # (A,)
+        log_stds = jnp.broadcast_to(log_stds, means.shape)  # (B, A)
```
Comment (zombie-einstein) [inline jumanji/training/networks/search_and_rescue/actor_critic.py:82]: I've not used Haiku a ton (usually Flax) so not sure if this is sensible, but need to have log-standard dev parameters broadcast over the input batch

File: jumanji/training/networks/distribution.py [hunk 3]
```diff
@@ -85,3 +85,42 @@ def kl_divergence(  # type: ignore[override]
         probs = jax.nn.softmax(self.logits)
         log_probs_other = jax.nn.log_softmax(other.logits)
         return jnp.sum(jnp.where(probs == 0, 0.0, probs * (log_probs - log_probs_other)), axis=-1)
+
+
+class NormalDistribution(Distribution):
```
[the remark this answers is not in the document]
Comment (zombie-einstein) [reply jumanji/training/networks/distribution.py:91]: Use distrax now

File: jumanji/training/networks/postprocessor.py [hunk 4]
```diff
@@ -47,6 +47,19 @@ def forward_log_det_jacobian(self, x: chex.Array) -> chex.Array:
...
+    """Identity Bijector for continuous distribution."""
+
+    def forward(self, x: chex.Array) -> chex.Array:
+        return x
+
+    def inverse(self, y: chex.Array) -> chex.Array:
+        return y
+
+    def forward_log_det_jacobian(self, x: chex.Array) -> chex.Array:
+        return jnp.zeros(x.shape[:-1])
```
Comment (zombie-einstein) [inline jumanji/training/networks/postprocessor.py:60]: I added this, as the log prob of the actions from the Gaussian is a single value (for each sample), this function seems to take a batch of actions as an input, so just return zeros in the shape of the batch here.

File: jumanji/training/networks/search_and_rescue/actor_critic.py (hunk 2 above)
Comment (sash-a) [reply jumanji/training/networks/search_and_rescue/actor_critic.py:82]: Honestly same, I was not involved in the training side of jumanji much. I am definitely a flax over haiku fan. This does look ok

**Review by sash-a — COMMENTED** [review]
How does it train, have you tried yet? 

I think main points are:
* Definitely should keep the action shape
* Need to double check if distrax will play nice with any vmapping during acting/learning
* Need to add a tanh bijector to squash the output between 0-1

File: requirements/requirements.txt [hunk 5]
```diff
@@ -1,4 +1,5 @@
 chex>=0.1.3
+distrax>=0.1.5
```
[the remark this answers is not in the document]
Comment (zombie-einstein) [reply requirements/requirements.txt:2]: Oh yeah of course, should have noticed that!

**Comment by zombie-einstein** [comment]
> How does it train, have you tried yet?
> 

It runs, but I actually need to leave it going, will try it later in the week.

File: jumanji/training/networks/distribution.py [hunk 6]
```diff
@@ -85,3 +86,27 @@ def kl_divergence(  # type: ignore[override]
...
+    """Normal distribution (with log standard deviations)."""
+
+    def __init__(self, means: chex.Array, log_stds: chex.Array):
+        self.dist = MultivariateNormalDiag(loc=means, scale_diag=jnp.exp(log_stds))
```
[the remark this answers is not in the document]
Comment (zombie-einstein) [reply jumanji/training/networks/distribution.py:95]: Oh I didn't realise, felt I'd used this ok in some other projects. You have a link to the issue?

File: jumanji/training/setup_train.py [hunk 7]
```diff
@@ -93,6 +94,10 @@ def _make_raw_env(cfg: DictConfig) -> Environment:
     if cfg.env.name in {"lbf", "connector"}:
         # Convert a multi-agent environment to a single-agent environment
         env = MultiToSingleWrapper(env)
+    if cfg.env.name == "search_and_rescue":
+        # Convert multi-agent search and rescue to single-agent, and
+        #  reshape actions into expected
+        env = MultiToSingleWrapper(env, action_shaper=lambda x: x.reshape(env.action_spec.shape))
```
[the remark this answers is not in the document]
Comment (zombie-einstein) [reply jumanji/training/setup_train.py:100]: > I'm not sure this is necessary, I think the shape should be fine if it's not flat. And if you need to do any reshaping I'd do it in the networks.

> What error do you get if it's not flat?

I need to try it out actually, figured this was the safer option and thought some of the code around the training distribution looked at dimensions of the actions, so was not sure about `[n-agents, n-actions]` being treated as multiple actions. I guess the only thing to check is how the Distribution treats an input, and treats an input `[n-batches, n-agents, n-action]` correctly/

File: jumanji/environments/swarms/search_and_rescue/utils.py [hunk 8]
```diff
@@ -84,3 +84,24 @@ def searcher_detect_targets(
...
+            boolean target detection flags.
+
+    Returns:
+        array of shape (n-agents,) contain
+    """
+    rewards = targets_found.astype(float)
+    norms = jnp.sum(rewards, axis=0)[jnp.newaxis]
+    rewards = jnp.where(norms > 0, rewards / norms, rewards)
+    rewards = jnp.sum(rewards, axis=1)
+    return rewards
```
[the remark this answers is not in the document]
Comment (zombie-einstein) [reply jumanji/environments/swarms/search_and_rescue/utils.py:107]: I had this in mind earlier, and then actually the change to carrying around the flags made it possible, I was thinking along the lines of this being a collective task, where the total score is the outcome (though agents may be working independently)

- It gives a clearer max possible score 
- If both agents can get a score, there is a edger-case/optima where they try their best to find a new agent in a new step, feels like it would incentivise clustering (or at least is a form of peverse incentive)
- It makes credit assignment a bit clearer

As such I felt the individual rewards could kind of end up decentivising co-operation.

Could make it optional, as  both cases are easy to code up. Feels like it kind of falls under rewards shaping that user might want to have some control over?

File: jumanji/training/networks/distribution.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (zombie-einstein) [reply jumanji/training/networks/distribution.py:95]: Oh yeah https://github.com/google-deepmind/distrax/issues/276

File: jumanji/environments/swarms/search_and_rescue/utils.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (sash-a) [reply jumanji/environments/swarms/search_and_rescue/utils.py:107]: > It gives a clearer max possible score
This is a very nice bonus

> It makes credit assignment a bit clearer
Not sure it does because if you give all agents that find each target 1 reward then credit is clear +1 if you find the target.

That being said there's pros and cons to both. If you have a look at connector it has a reward file which you could copy and make both options - similar to how you currently do the generator.

File: jumanji/training/networks/search_and_rescue/actor_critic.py [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+        are reshaped by the wrapped environment).
+
+    Args:
+        search_and_rescue: `SearchAndRescue` environment.
+        layers: List of hidden layer dimensions.
+
+    Returns:
+        Continuous action space MLP action and critic networks.
+    """
+    n_actions = prod(search_and_rescue.action_spec.shape)
```
[the remark this answers is not in the document]
Comment (sash-a) [reply jumanji/training/networks/search_and_rescue/actor_critic.py:51]: A good example of this in in lbf networks as this is also a multi-agent env. You can see it uses a `MultiCategoricalParametricDistribution` so basically treating it as a multi-discrete problem. So you'll see a lot of the outputs are of shape `(B, N, ...)` (B=batch, N=num agents).

File: jumanji/training/networks/distribution.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (sash-a) [reply jumanji/training/networks/distribution.py:95]: Indeed, I'll have a look tomorrow to try work out if this will be an issue or not.

**Comment by sash-a** [comment]
If this becomes too much of a hassle to test I think testing it mava will be much easier - all the networks etc are set up it would just require a small wrapper. See if you can get some signal here but if not let's switch to mava and see if that is easier to set up

File: jumanji/training/setup_train.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (zombie-einstein) [reply jumanji/training/setup_train.py:100]: Oh great I'll grab a look

File: jumanji/environments/swarms/search_and_rescue/utils.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (zombie-einstein) [reply jumanji/environments/swarms/search_and_rescue/utils.py:107]: Sounds like a plan, it'd be good then to have this convert between shared and individual rewards

File: jumanji/training/networks/distribution.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (zombie-einstein) [reply jumanji/training/networks/distribution.py:95]: Is the issue just shape related (i.e. it's not a nnumerical problem) as these changes do compile and run Ok, so at least does not seem to be hitting this issues

[the remark this answers is not in the document]
Comment (sash-a) [reply jumanji/training/networks/distribution.py:95]: I think it should error if there is an issue. At least that's what the readme implies

> complex objects such as distrax.MultivariateNormalDiag may generate exceptions when used as inputs or outputs of a vmap-ed or pmap-ed function.

**Comment by zombie-einstein** [comment]
So I implemented the the continuos actor and critic to return `[batch, n-agents, n-actions]` and `[batch, n-agents]` respectively, and it raises an error 

```
Error in rank compatibility check: input 0 has rank 2 (shape (20, 2)) but expected 1; input 3 has rank 2 (shape (20, 2)) but expected 1.
```

raised here by rlax when calculating the advantage [here](https://github.com/instadeepai/jumanji/blob/1ec880828a6bc72e31f8343f0dd4c924b506d945/jumanji/training/agents/a2c/a2c_agent.py#L128), I think because the critic is returning a rank matrix of values per agents. 

Comparing to the lbf policy, though it generated actions for multiple agents, it only returns a single values for all the agents by summing over the embedding [here](https://github.com/instadeepai/jumanji/blob/1ec880828a6bc72e31f8343f0dd4c924b506d945/jumanji/training/networks/lbf/actor_critic.py#L132) and then it returns individual actions per agent, but then when calculating the log-prob, it sums over the agent dimension [here](https://github.com/instadeepai/jumanji/blob/1ec880828a6bc72e31f8343f0dd4c924b506d945/jumanji/training/networks/parametric_distribution.py#L94) as this value is set to one, to get a an array `[batch-size,]` of log-probs.

We can probably do the same for the log-prob calculation for the continuos actions (i.e. sum over the agent dimensions) , but I'm a little less sure of the equivalent  single value across agent actions. Looks like this needs to be done in the critic network, guess I could flatten the individual observations into one large one?

**Comment by sash-a** [comment]
Sorry I guess I wasn't clear, my comments were for the actor network. The critic should return a single value over (what we call in MARL) the joint-observation (ideally the state, but we don't have access to a true state representation here and the joint-observation is close enough). So you can basically pass the observation concatenated on the agent dim to the critic and return a single value for each item in the batch so the shape should be `batch`. This should solve the GAE issue as your value will now be the expected shape. 

It should also be fine for your actor to return `[batch, n-agents, n-actions]` as it should be what the environments expects

**Comment by zombie-einstein** [comment]
> Sorry I guess I wasn't clear, my comments were for the actor network. The critic should return a single value over (what we call in MARL) the joint-observation (ideally the state, but we don't have access to a true state representation here and the joint-observation is close enough). So you can basically pass the observation concatenated on the agent dim to the critic and return a single value for each item in the batch so the output shape of the values should be `(batch,)` which should match other environments (I think). This should solve the GAE issue as your value will now be the expected shape.
> 
> It should also be fine for your actor to return `[batch, n-agents, n-actions]` as it should be what the environments expects

Thanks yeah just made these changes, need to finalise the bijector, but seems to be working ok now!

**Comment by zombie-einstein** [comment]
Hi @sash-a, so this is running, and seems to be learning, something though not clear-cut. 

This is the random policy  (blue) v a2c stochastic eval (orange) with 100 targets and 2 searchers

image

so it's doing marginally better but not by much. The greedy policy seems to settle on a consistently better strategy, but again not by a huge amount, and seems to not continue improving

image

Though in the current setup it's maybe hard to see what an optimal strategy could be, if the targets are distributed uniformly then a random search is probably not bad as long as you also try avoid other searchers. I image if the agents had some memory or were recurrent networks then they could learn a search pattern?

FYI I was going to merge this as it was meant to be a short branch just to sort out the training functionality.

File: jumanji/training/networks/parametric_distribution.py [hunk 10]
```diff
@@ -173,3 +177,17 @@ def __init__(self, action_spec_num_values: chex.ArrayNumpy):
...
     def create_dist(self, parameters: chex.Array) -> CategoricalDistribution:
         return CategoricalDistribution(logits=parameters)
+
+
+class ContinuousActionSpaceNormalDistribution(ParametricDistribution):
+    """Normal distribution for continuous action spaces"""
+
+    def __init__(self, n_actions: int):
+        super().__init__(
+            param_size=n_actions,
```
[the remark this answers is not in the document]
Comment (zombie-einstein) [reply jumanji/training/networks/parametric_distribution.py:188]: Looking into it, I'm not sure it's accessed so think this is ok

**Comment by zombie-einstein** [comment]
> What is the current default setup? 

So this is random uniformly distributed targets, with random noisey movements (though checking now this is maybe a bit too noisy), and the rewards are summed over the agents (via the wrapper), with rewards shared in the case targets are simultaneously detected. Agents see other searching agents (within range) as part of their observation. So apart from other agents location they don't have a lot of info to go off.

One thing we discussed was having the agents be able to visualise targets as part of their observation, feels like it would make it easier, for example as agents could look at areas of low found target density (at least for a uniform distribution). Can see how it could be an optional feature, though you mentioned maybe tricky with the observation spec

> I think I should have some time (possibly tomorrow) to test this in Mava I imagine some centralized training and memory would help here like you pointed out and that will be difficult to add here but already exists in Mava. Unless you want to test it in Mava, but I am happy to as I realize this would be another library to learn for you 😄

If you have time that'd be great, but let me know if not, I was keen to try out Mava anyway and have a bit of time.

**Comment by sash-a** [comment]
> though checking now this is maybe a bit too noisy

One important thing to check is that targets are significantly slower than the searchers, maybe 50% of their max speed, just so they are catchable.

I think test out whatever you think might make the problem easier. Number 1 thing I'd recommend is less agents and targets.


> If you have time that'd be great, but let me know if not, I was keen to try out Mava anyway and have a bit of time.

I'll try to get to it tomorrow, at least I should be able to set up a branch where it will be piping through, not sure if I'll be able to run thorough tests

**Comment by zombie-einstein** [comment]
>> FYI I was going to merge this as it was meant to be a short branch just to sort out the training functionality.
> No stress on this taking longer, unfortunately it always does take longer than you'd expect, I think let's keep it here for now as it's definitely easier.

Sorry more mean that I wanted to merge this branch into the main fork, just so it's a bit easier to track changes with new branches, then this one long-lived branch of the fork!

**Comment by sash-a** [comment]
Ah I see, good plan easier to track new changes :smile: 

Also I briefly started but didn't finish the Mava stuff on Friday, should finish it today :smile:
